File: js/src/table.js
(function (global) {
	
	var EThing = global.EThing;
	
	
	
	/**
	 * Constructs a Table instance from an object decribing a table. Should not be called directly. Use instead {@link EThing.list}.
	 * @protected
	 * @class The Table resource handle data in rows and columns
	 * @memberof EThing
	 * @extends EThing.Resource
	 * @param {object} json
	 */
	EThing.Table = function(json)
	{
		EThing.Resource.call(this, json);
	}
	EThing.utils.inherits(EThing.Table, EThing.Resource);
	
	
	// specific methods
	/**
	 * The number of rows in this table
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @returns {number}
	 */
	EThing.Table.prototype.length = function(){
		return this._json['length'];
	}
	/**
	 * The maximum number of rows allowed in this table. Returns null if this feature is disable (number of rows is unlimited).
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @returns {number|null}
	 */
	EThing.Table.prototype.maxLength = function(){
		return this._json.maxLength;
	}
	/**
	 * Returns the amount of seconds after which a __row__ is automatically removed, or null if this feature is not enable.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @returns {number}
	 */
	EThing.Table.prototype.expireAfter = function(){
		return this._json.expireAfter;
	}
	/**
	 * Returns the keys in this table. __The default keys ("id" and "date" are not listed)__.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @returns {string[]}
	 */
	EThing.Table.prototype.keys = function(){
		var keys = [];
		for(var k in this._json.keys)
			if(this._json.keys.hasOwnProperty(k))
				keys.push(k);
		return keys;
	}
	/**
	 * Last time the content of this resource was modified
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @returns {Date}
	 */
	EThing.Table.prototype.contentModifiedDate = function() {
		return new Date(this._json.contentModifiedDate);
	}
	/**
	 * Returns rows.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @param {Object} [options] Customize the selection
	 * @param {number} [options.start=0] Position of the first rows to return. If start is negative, it starts from the end.
	 * @param {number} [options.length] Maximum number of rows to return.
	 * @param {string} [options.sort] The key on which to do the sorting, by default the sort is made by date.
	 * @param {string} [options.query] A query string to filter the results.
	 * @param {string} [options.fields] Fields of the results to return. If omitted, all the fields are returned.
	 * @param {string} [options.datefmt] the format of the date field (values: "RFC3339"(default), "TIMESTAMP", "TIMESTAMP_MS", "ISO8601", "RSS").
	 * @param {function(data,XHR,options)} [callback] It is executed once the request is complete whether in failure or success
	 * @returns {EThing.Table} The instance on which this method was called.
	 * @example
	 * // returns all the content of a table :
	 * table.select().done(function(data){
	 *   // success
	 * });
	 *
	 * // returns the last 10 rows sorted by the "foo" column :
	 * table.select({start: -10, sort: "foo"}).done(function(data){
	 *   // success
	 * });
	 *
	 */
	EThing.Table.prototype.select = function(options,callback){
		var args = [].slice.call(arguments);
		return this.deferred(function(){
				args.unshift(this);
				return EThing.Table.select.apply(EThing, args);
			});
	}
	/**
	 * Performs statistics on a specific column.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @param {string} key The column name on which the statistics should be performed.
	 * @param {string} [query] A query string to filter the data.
	 * @param {function(statistics,XHR,options)} [callback] It is executed once the request is complete whether in failure or success
	 * @returns {Object} The instance on which this method was called.
	 *
	 */
	EThing.Table.prototype.computeStatistics = function(key,query,callback){
		var args = [].slice.call(arguments);
		return this.deferred(function(){
				args.unshift(this);
				return EThing.Table.computeStatistics.apply(EThing, args);
			});
	}
	/**
	 * Removes one or multiple rows.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @param {string|string[]} id The id of the row(s) to be removed
	 * @param {function(data,XHR,options)} [callback] it is executed once the request is complete whether in failure or success
	 * @returns {EThing.Table} The instance on which this method was called.
	 * @example
	 * // removes the last 10 rows :
	 * table.select(-10).done(function(data){
	 *   this.removeRow(data.map(function(row){
	 *     return row.id;
	 *   }));
	 * });
	 *
	 */
	EThing.Table.prototype.removeRow = function(id,callback){
		var args = [].slice.call(arguments);
		return this.deferred(function(){
				args.unshift(this);
				return EThing.Table.removeRow.apply(EThing, args);
			});
	}
	/**
	 * Set new data to a row.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @param {object} data the data to be updated, it must contain the row id.
	 * @param {function(data,XHR,options)} [callback] it is executed once the request is complete whether in failure or success
	 * @returns {EThing.Table} The instance on which this method was called.
	 * @example
	 * table.replaceRow({
	 * 	 'id': 'g45Tdk4',
	 * 	 'value': 45
	 * }).done(function(data){
	 *   // updated !
	 * });
	 *
	 */
	EThing.Table.prototype.replaceRow = function(data,callback){
		var args = [].slice.call(arguments);
		return this.deferred(function(){
				args.unshift(this);
				return EThing.Table.replaceRow.apply(EThing, args);
			});
	}
	
	/**
	 * Finds a single row matching the query and replaces it.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @param {string} query A query string to filter the data.
	 * @param {object} data the new data.
	 * @param {boolean} [upsert] If true, the data will be inserted if no match is found. (default to false).
	 * @param {function(data,XHR,options)} [callback] it is executed once the request is complete whether in failure or success
	 * @returns {EThing.Table} The instance on which this method was called.
	 * @example
	 * table.findOneAndReplace("name=='foo'", {
	 * 	 'name': 'foo',
	 * 	 'value': 'bar'
	 * }).done(function(table){
	 *   // updated or inserted if not found !
	 * });
	 *
	 */
	EThing.Table.prototype.findOneAndReplace = function(query,data,upsert,callback){
		var args = [].slice.call(arguments);
		return this.deferred(function(){
				args.unshift(this);
				return EThing.Table.findOneAndReplace.apply(EThing, args);
			});
	}
	
	/**
	 * Insert new data into the table. The data argument must be a key/value object.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @param {object} data
	 * @param {string} [invalid_field] The behaviour to adopt when an invalid field name appears. The value must be one of the following : "rename","stop","skip","none".
	 * @param {function(data,XHR,options)} [callback] it is executed once the request is complete whether in failure or success
	 * @returns {EThing.Table} The instance on which this method was called.
	 * @example
	 * table.insert({
	 *   'field1': "foobar",
	 *   'field2': 3.14,
	 *   'field3': true
	 * }).done(function(){
	 *   // success
	 * });
	 *
	 */
	EThing.Table.prototype.insert = function(data, callback){
		var args = [].slice.call(arguments);
		return this.deferred(function(){
			args.unshift(this);
			return EThing.Table.insert.apply(this,args);
		});
	}
	
	/**
	 * Replace the content of this table by a new set of data.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @param {object[]} data
	 * @param {string} [invalid_field] The behaviour to adopt when an invalid field name appears. The value must be one of the following : "rename","stop","skip","none".
	 * @param {bool} [skip_error] Whether to skip data on error or not
	 * @param {function(data,XHR,options)} [callback] it is executed once the request is complete whether in failure or success
	 * @returns {EThing.Table} The instance on which this method was called.
	 * // copy table content
	 * var tableSrc, tableDst;
	 * tableSrc.select().done(function(data){
	 *   tableDst.import(data);
	 * });
	 *
	 */
	EThing.Table.prototype.import = function(data, callback){
		var args = [].slice.call(arguments);
		return this.deferred(function(){
			args.unshift(this);
			return EThing.Table.import.apply(this,args);
		});
	}
	
	/**
	 * Returns the link to access the content.
	 * @memberof EThing.Table
	 * @this {EThing.Table}
	 * @param {boolean} [auth=false] wether or not attach any authentication element. Necessary if you are not using {@link EThing.request}.
	 * @returns {string}
	 * @example
	 * // using EThing.request() :
	 * EThing.request(table.getContentUrl()).done(function(rows){
	 *   // success, rows is an array of object 
	 *   console.log('number of rows : '+rows.length);
	 * });
	 * 
	 * // or using jQuery :
	 * $.getJSON(table.getContentUrl(true)).done(function(rows){
	 *   // success
	 * });
	 */
	EThing.Table.prototype.getContentUrl = function(auth) {
		return EThing.toApiUrl('tables/'+this.id(),auth);
	}
	
	
	
	
	/**
	 * Creates a new Table from the following attributes :
	 *   - name {string} __ required__ the name of the table
	 *   - description {string} a string describing this table 
	 *   - data {object} key/value pairs to attach to this table
	 *   - expireAfter {number} amount of seconds after which a row is automatically removed, 0 means unlimited. Default to 0.
	 *   - maxLength {number} the maximum number of rows allowed in this table. 0 means unlimited. Default to 5000.
	 *
	 * @method EThing.Table.create
	 * @param {object} attributes
	 * @param {function(data,XHR,options)} [callback] it is executed once the request is complete whether in failure or success
	 * @returns {Deferred} a {@link http://api.jquery.com/category/deferred-object/|jQuery like Promise object}. {@link EThing.request|More ...} 
	 * @fires EThing#ething.table.created
	 * @example
	 * // get a resource by its id
	 * EThing.Table.create({
	 *   name: "foobar",
	 *   expireAfter: 3600*24*7 // the data are automatically removed after 7 weeks
	 * }).done(function(resource){
	 *     console.log('table created : ' + resource.name());
	 * })
	 */
	EThing.Table.create = function(a,callback){
		
		if(typeof a == "string")
			a = {
				'name': a
			};

		return EThing.request({
			'url': '/tables',
			'dataType': 'json',
			'method': 'POST',
			'contentType': "application/json; charset=utf-8",
			'data': a,
			'converter': resourceConverter
		},callback).done(function(r){
			EThing.trigger('ething.table.created',[r]);
		});
		
	};
	
	
	EThing.Table.select = function(a,options,callback){
	
		var table_id = null, context;
		if(a instanceof EThing.Table){
			context = a;
			table_id = a.id();
		}
		else if(isResourceId(a))
			table_id = a;
		else {
			throw "First argument must be a Table object or a table id !";
			return;
		}
		
		if((typeof callback == 'undefined') && (typeof options == 'function')){
			callback = options;
			options = null;
		}
		
		options = options || {};
		
		if(Array.isArray(options.fields)){
			options.fields = options.fields.join(',');
		}
		
		
		return EThing.request({
			'url': '/tables/' + table_id + '?' + EThing.utils.param({'start':options.start,'length':options.length,'sort':options.sort,'q':options.query,'fields':options.fields,'datefmt':options.datefmt}),
			'method': 'GET',
			'dataType': 'json',
			'context': context
		},callback);
		
	}
	
	EThing.Table.computeStatistics = function(a,key,query,callback){
	
		var table_id = null, context;
		if(a instanceof EThing.Table){
			context = a;
			table_id = a.id();
		}
		else if(isResourceId(a))
			table_id = a;
		else {
			throw "First argument must be a Table object or a table id !";
			return;
		}
		
		if((typeof callback == 'undefined') && (typeof query == 'function')){
			callback = query;
			query = null;
		}
		
		
		return EThing.request({
			'url': '/tables/' + table_id + '/statistics?' + EThing.utils.param({'key':key,'q':query}),
			'method': 'GET',
			'dataType': 'json',
			'context': context
		},callback);
		
	}
	
	
	/*
	Table,id,callback
	*/
	EThing.Table.removeRow = function(a,b,c){
	
		var table_id = null, context;
		if(a instanceof EThing.Table){
			context = a;
			table_id = a.id();
		}
		else if(isResourceId(a))
			table_id = a;
		else {
			throw "First argument must be a Table object or a table id !";
			return;
		}
		
		var id = Array.isArray(b) ? b : [b];
		var callback = c;
		
		return EThing.request({
			'url': '/tables/' + table_id + '/remove',
			'dataType': 'json',
			'method': 'POST',
			'contentType': "application/json; charset=utf-8",
			'data': id,
			'context': context,
			'converter': resourceConverter
		},callback);
		
	}
	
	/*
	Table,data,callback
	*/
	EThing.Table.replaceRow = function(a,b,c){
	
		var table_id = null, context;
		if(a instanceof EThing.Table){
			context = a;
			table_id = a.id();
		}
		else if(isResourceId(a))
			table_id = a;
		else {
			throw "First argument must be a Table object or a table id !";
			return;
		}
		
		if(typeof b === 'object' && !isResourceId(b.id))
			throw "Second argument must be an object containing at least a document id";
		
		var docId = b.id
		var callback = c;
		
		return EThing.request({
			'url': '/tables/' + table_id + '/id/'+docId,
			'dataType': 'json',
			'method': 'POST',
			'contentType': "application/json; charset=utf-8",
			'data': b,
			'context': context,
			'headers': {
				"X-HTTP-Method-Override": "PATCH"
			}
		},callback);
		
	}
	
	/*
	Table,query,data[,upsert][,callback]
	*/
	EThing.Table.findOneAndReplace = function(a,b,c,d,e){
	
		var table_id = null, context, callback = e, upsert = false;
		if(a instanceof EThing.Table){
			context = a;
			table_id = a.id();
		}
		else if(isResourceId(a))
			table_id = a;
		else {
			throw "First argument must be a Table object or a table id !";
			return;
		}
		
		if(typeof b !== 'string')
			throw "Second argument must be a query string";
		
		if(typeof c !== 'object')
			throw "Third argument must be an object";
		
		if(typeof e === 'undefined' && typeof d === 'function'){
			callback = d;
			upsert = false;
		}
		if(typeof d === 'boolean') {
			upsert = d;
		}
		
		return EThing.request({
			'url': '/tables/' + table_id + '/replace?' + EThing.utils.param({'q':b, 'upsert':upsert}),
			'dataType': 'json',
			'method': 'POST',
			'contentType': "application/json; charset=utf-8",
			'data': c,
			'context': context,
			'converter': resourceConverter
		},callback);
		
	}
	
	/*
	Table,kv,callback
	Table,column,data,callback
	
	kv = {
		...
		columnX: valueX,
		...
	}
	
	Note : if valueX is null, so no data will be appended and the columnX will be created if it does not exist
	
	*/
	EThing.Table.insert = function(a,postData,c,d){
		
		var callback, table_id = null, context, invalid_field;
		
		if(a instanceof EThing.Table){
			context = a;
			table_id = a.id();
		}
		else if(isResourceId(a))
			table_id = a;
		else {
			throw "First argument must be a Table object or a table id !";
			return;
		}
		
		if( typeof c == 'string' ){
			invalid_field = c;
			callback = d;
		}
		else {
			callback = c;
		}
		
		
		return EThing.request({
			'url': '/tables/' + table_id + '?' + EThing.utils.param({'invalid_field':invalid_field}),
			'dataType': 'json',
			'method': 'POST',
			'contentType': "application/json; charset=utf-8",
			'data': postData,
			'context': context,
			'converter': resourceConverter
		},callback);
		
	}
	
	
	EThing.Table.import = function(table,data){
		var context,invalid_field,skip_error,callback;
		
		if(table instanceof EThing.Table){
			context = table;
			table = table.id();
		}
		else if(!isResourceId(table)){
			throw "First argument must be a Table object or a table id !";
			return;
		}
		
		if(!Array.isArray(data))
			throw "The data must be an array of objects";
		
		for(var i=2; i<arguments.length; i++){
			switch(typeof arguments[i]){
				case 'function':
					callback = arguments[i];
					break;
				case 'boolean':
					skip_error = arguments[i];
					break;
				case 'string':
					invalid_field = arguments[i];
					break;
			}
		}
		
		return EThing.request({
			'url': '/tables/' + table + '?' + EThing.utils.param({'skip_error':skip_error,'invalid_field':invalid_field}),
			'dataType': 'json',
			'method': 'PUT',
			'contentType': "application/json; charset=utf-8",
			'data': data,
			'context': context,
			'converter': resourceConverter
		},callback);
		
	}
	
	
	
})(this);

File: ething/webserver/routes/table.py
# coding: utf-8
from future.utils import string_types

from flask import request, Response
from ..server_utils import *
import csv
from ething.core.ShortId import ShortId


def install(core, app, auth, **kwargs):

    @app.route('/api/tables', methods=['POST'])
    @auth.required('table:write resource:write')
    def tables():
        """
        ---
        post:
          tags:
            - table
          description: |-
            Creates a new table.

            You may want to pass directly the content of the table in the same request. To do so, just pass the data through the key 'content' of the metadata object;

            example:

            ```json
            {
               "name": "foobar.db",
               "content": [
                    {
                        "temperature": 12.5,
                        "pressure": 101325
                    }
               ]
            }
            ```

            #### cURL example

            The next command will create a new table 'mytable.db'.

            ```bash
            curl
                -H 'X-API-KEY: <YOUR_API_KEY>'
                -H "Content-Type: application/json"
                -X POST
                -d '{"name":"mytable.db"}'
                http://localhost:8000/api/tables
            ```

            If the command was successful, a response containing the meta data of the created table will be given back.
            You will find in it the id of that table.
            This id is a unique string identifying this table and is necessary to make any operation on it.

            ```json
            {
              "id":"56_df0f",
              "name":"mytable.db",
              "data":null,
              "description":null,
              "maxLength":null,
              "expireAfter":null,
              "type":"Table",
              "createdBy":null,
              "createdDate":"2016-02-12T14:49:30+00:00",
              "modifiedDate":"2016-02-15T13:03:20+00:00",
              "length":421,
              "keys":{
                 "temp1":421,
                 "temp2":421
              },
              "location":null
            }
            ```
          parameters:
            - name: metadata
              in: body
              description: |-


                The metadata of the table to be created.

                example:

                ```json
                {
                    "name":"mytable.db"
                }
                ```

              required: true
              schema:
                $ref: '#/definitions/Table'
          responses:
            '200':
              description: The table was successfully created
              schema:
                $ref: '#/definitions/Table'
        """
        attr = request.get_json()

        if isinstance(attr, dict):

            content = None

            if 'content' in attr:
                content = attr['content']
                attr.pop('content')

            attr.setdefault('createdBy', g.auth.resource)

            r = core.create('resources/Table', attr)

            if r:

                if content:
                    r.importData(content)

                response = app.jsonify(r)
                response.status_code = 201
                return response
            else:
                raise Exception('Unable to create the table')

        raise Exception('Invalid request')

    table_fmts = ["json", "json_pretty", "csv", "csv_no_header"]
    date_fmts = ['timestamp', 'timestamp_ms', 'rfc3339']
    table_get_args = {
        'start': fields.Int(missing=0, description="Position of the first rows to return. If start is negative, the position will start from the end. (default to 0)"),
        'length': fields.Int(validate=validate.Range(min=0), description="Maximum number of rows to return. If not set, returns until the end"),
        'sort': fields.Str(description='the key on which to do the sorting, by default the sort is made by date ascending. To make the sort descending, prepend the field name by minus "-". For instance, "-date" will sort by date descending'),
        'query': fields.Str(data_key="q", load_from="q", description="Query string for filtering results"),
        'date_format': fields.Str(data_key="datefmt", load_from="datefmt", validate=lambda val: val.lower() in date_fmts, missing='rfc3339', description="the format of the date field (default to RFC3339) : %s" % ','.join(date_fmts)),
        'fmt': fields.Str(validate=validate.OneOf(table_fmts), missing='json', description="the output format (default to JSON) : %s" % ','.join(table_fmts)),
    }

    table_put_args = {
        'invalid_field': fields.Str(validate=validate.OneOf(['rename', 'stop', 'skip', 'none']), missing='rename', description="The behaviour to adopt when an invalid field name appears."),
        'skip_error': fields.Bool(missing=True, description="Whether to skip data on error or not."),
    }

    table_post_args = {
        'invalid_field': fields.Str(validate=validate.OneOf(['rename', 'stop', 'skip', 'none']), missing='rename', description="The behaviour to adopt when an invalid field name appears."),
    }

    @app.route('/api/tables/<id>', methods=['GET', 'PUT', 'POST'])
    @use_multi_args(GET=table_get_args, PUT=(table_put_args, ('query',)), POST=(table_post_args, ('query',)))
    @auth.required(GET='table:read resource:read', PUT='table:write resource:write', POST='table:write table:append resource:write')
    def table(args, id):
        """
        ---
        get:
          tags:
            - table
          description: |-
            Retrieves the content of a table.

            #### cURL examples

            ```bash
            # get all the data of a table :
            curl -H 'X-API-KEY: <YOUR_API_KEY>' http://localhost:8000/api/tables/<TABLE_ID>

            # only the first 20 rows :
            curl -H 'X-API-KEY: <YOUR_API_KEY>' http://localhost:8000/api/tables/<TABLE_ID>?start=0&length=20

            # only the last 20 rows :
            curl -H 'X-API-KEY: <YOUR_API_KEY>' http://localhost:8000/api/tables/<TABLE_ID>?start=-20

            # only the last 10 rows sorted by the field "temperature" in ascending order
            # (put a minus before the name of the field if you want to sort in descending order)
            curl -H 'X-API-KEY: <YOUR_API_KEY>' http://localhost:8000/api/tables/<TABLE_ID>?start=-10&sort=temperature
            ```
          produces:
            - application/json
            - text/plain
          responses:
            '200':
              description: The records of this table
              schema:
                type: array
                items:
                  type: object
                  description: record's object. Every record has at least the 'id' and 'date' keys.
                  properties:
                    id:
                      type: string
                      description: an unique id to identify a record
                    date:
                      type: string
                      format: date-time
                      description: the create date of this record
        put:
          tags:
            - table
          description: Set the content of a table. The new data will erase the previous one.
          consumes:
            - application/json
          parameters:
            - name: content
              in: body
              description: |-
                The content to be inserted as an array of object.

                The data must be sent in a JSON formatted object :

                ```json
                [{
                    "<KEY>":<VALUE>
                }]
                ```

                example:

                ```json
                [
                    {
                      "date": "2016-02-06T15:03:07+01:00",
                      "temperature": 12.5,
                      "pressure": 101325
                    },
                    {
                      "date": "2016-02-06T16:03:07+01:00",
                      "temperature": 13.5,
                      "pressure": 101212
                    }
                ]
                ```

                If the 'date' field is not present, the current date will be set automatically.
                If an 'id' field is present, it will be automatically be resetted to a new value.


              required: true
              schema:
                type: array
                items:
                  type: object
          responses:
            '200':
              description: The content was successfully set. The table metadata is returned.
              schema:
                $ref: '#/definitions/Table'
        post:
          tags:
            - table
          description: Insert a new record in a table
          parameters:
            - name: record
              in: body
              description: |-
                The record to be inserted.

                The data must be sent in a JSON formatted object :

                ```json
                {
                    "<KEY>":<VALUE>
                }
                ```

                cURL example :

                ```bash
                curl
                    -H 'X-API-KEY: <YOUR_API_KEY>'
                    -H "Content-Type: application/json"
                    -X POST
                    -d '{"temperature":15.2, "comment":"outdoor"}'
                    http://localhost:8000/api/tables/<TABLE_ID>
                ```



              required: true
              schema:
                type: object
          responses:
            '200':
              description: The record was successfully inserted. The table metadata is returned.
              schema:
                $ref: '#/definitions/Table'
        """

        r = app.getResource(id, ['Table'])

        if request.method == 'GET':

            fmt = args.pop('fmt').lower()

            if fmt == "json":
                return app.jsonify(r.select(**args))
            elif fmt == "json_pretty":
                return app.jsonify(r.select(**args), indent=4)
            elif fmt == "csv" or fmt == "csv_no_header":
                args['show_header'] = (fmt == "csv")

                fields = request.args.get('fields')
                if fields is not None:
                    fields = fields.replace(' ', ',').replace(
                        ';', ',').replace('|', ',').split(',')
                    args['fields'] = fields

                return Response(r.writeCSV(**args), mimetype='text/csv')

            raise Exception('Invalid request')

        elif request.method == 'PUT':

            data = request.get_json()

            if data:
                r.importData(data, args['invalid_field'], args['skip_error'])
                return app.jsonify(r)
            else:
                raise Exception('No data.')

        elif request.method == 'POST':

            data = request.get_json()

            if data:
                r.insert(data, args['invalid_field'])
                return app.jsonify(r)
            else:
                raise Exception('No data.')

    table_action_remove_args = {
        'ids': fields.DelimitedList(fields.Str(), missing=[], description="The records id to be removed as a comma separated list."),
    }

    @app.route('/api/tables/<id>/remove', methods=['POST'])
    @use_args(table_action_remove_args, locations=('query', 'form'))
    @auth.required('table:write resource:write')
    def table_delete_rows(args, id):
        """
        ---
        post:
          tags:
            - table
          description: Remove one or more records in a table
          responses:
            '200':
              description: The records was successfully deleted
              schema:
                $ref: '#/definitions/Table'
        """
        r = app.getResource(id, ['Table'])
        ids = args['ids'] or []

        if request.json:
            more_ids = request.json

            if isinstance(more_ids, string_types):
                more_ids = [more_ids]

            if not isinstance(more_ids, list):
                raise Exception('The key ids must be an array of record id.')

            ids = ids + more_ids

        if len(ids) > 0:
            for id in ids:
                if not ShortId.validate(id):
                    raise Exception('Must be an array of record id.')

            nb = r.remove_rows(ids)

            if nb == len(ids):
                # all the specified documents/rows were removed
                return app.jsonify(r)
            else:
                # all or only certain documents/rows could not have been removed
                raise Exception(
                    'Some or all documents could not have been removed.')

    table_action_replace_args = {
        'q': fields.Str(required=True, description="A query that select the rows to update"),
        'invalid_field': fields.Str(validate=validate.OneOf(['rename', 'stop', 'skip', 'none']), missing='rename', description="The behaviour to adopt when an invalid field name appears."),
        'upsert': fields.Bool(missing=False, description="If true and no records was found, the data will be added to the table as a new record."),
    }

    @app.route('/api/tables/<id>/replace', methods=['POST'])
    @use_args(table_action_replace_args, locations=('query',))
    @auth.required('table:write resource:write')
    def table_replace_rows(args, id):
        """
        ---
        post:
          tags:
            - table
          description: Update records in a table
          responses:
            '200':
              description: The records was successfully updated
              schema:
                $ref: '#/definitions/Table'
        """

        r = app.getResource(id, ['Table'])
        data = request.get_json()

        if data:

            r.replaceRow(args['q'], data, args['invalid_field'],
                         args['upsert'])
            return app.jsonify(r)
        else:
            raise Exception('No data.')

    table_statistics_args = {
        'q': fields.Str(description="A query string to select the rows used for the statistics computation", missing=None),
        'key': fields.Str(required=True, description="the name of the key. Statistics can only be computed for a single key."),
    }

    @app.route('/api/tables/<id>/statistics')
    @auth.required('table:read resource:read')
    @use_args(table_statistics_args)
    def table_statistics(args, id):
        """
        ---
        get:
          tags:
            - table
          description: Compute statistics of a column (=key)
          responses:
            '200':
              description: The records was successfully updated
              schema:
                type: object
                description: The statistics object.
        """
        r = app.getResource(id, ['Table'])
        return app.jsonify(r.computeStatistics(args['key'], args['q']))

    table_cell_id_patch_args = {
        'invalid_field': fields.Str(validate=validate.OneOf(['rename', 'stop', 'skip', 'none']), missing='rename'),
    }

    @app.route('/api/tables/<id>/id/<doc_id>', methods=['GET', 'DELETE', 'PATCH'])
    @use_multi_args(PATCH=(table_cell_id_patch_args, ('query',)))
    @auth.required(GET='table:read resource:read', DELETE='table:write resource:write', PATCH='table:write resource:write')
    def table_cell_id(args, id, doc_id):

        r = app.getResource(id, ['Table'])

        if request.method == 'GET':

            doc = r.getRow(doc_id)

            if doc is None:
                raise Exception(
                    'The document with id=%s does not exist.' % doc_id)

            return app.jsonify(doc)

        elif request.method == 'DELETE':
            r.remove_row(doc_id)
            return '', 204

        elif request.method == 'PATCH':

            data = request.get_json()

            if data:
                doc = r.replaceRowById(doc_id, data, args['invalid_field'])
                if doc:
                    return app.jsonify(doc)
                else:
                    raise Exception(
                        'The document with id=%s does not exist.' % doc_id)
            else:
                raise Exception('No data')
